ts/Core/Chart/Chart.ts

```diff
@@ -2405,20 +2403,80 @@ class Chart {
         const tempWidth = chart.plotWidth;
 
         axes.some(function (axis: Axis): (boolean|undefined) {
+            // Left side must be align: right and right side must
+            // have align: left for labels
+            axis.reserveSpaceDefault = (
+                axis.side === 0 ||
+                axis.side === 2 ||
+                { 1: 'left', 3: 'right' }[axis.side] === axis.labelAlign
+            );
+
+            // Set the explicit or automatic label alignment
+            axis.labelAlign = axis.options.labels.align ||
+                axis.autoLabelAlign(axis.labelRotation as any);
+
             if (
                 axis.horiz &&
                 axis.visible &&
                 axis.options.labels.enabled &&
                 axis.series.length
             ) {
-                // 21 is the most common correction for X axis labels
-                correction = 21;
-                return true;
+                // Create the axisGroup and gridGroup elements on first
+                // iteration
+                if (!axis.axisGroup) {
+                    if (axis.coll === 'colorAxis') {
+                        axis.axisParent = axis.legendItem?.group;
+                    }
+
+                    axis.gridGroup = axis.createGroup(
+                        'grid',
+                        '-grid',
+                        axis.options.gridZIndex
+                    );
+                    axis.axisGroup = axis.createGroup(
+                        'axis',
+                        '',
+                        axis.options.zIndex
+                    );
+                    axis.labelGroup = axis.createGroup(
+                        'axis-labels',
+                        '-labels',
+                        axis.options.labels.zIndex
+                    );
```

Comment:
This code block is in principle identical to this:

https://github.com/highcharts/highcharts/blob/c286688e549da6aa1080ba649d9b1af7d16b54b6/ts/Core/Axis/Axis.ts#L3478-L3494

It should be refactored out so that the `Axis.createGroup` function takes care of adding all the groups, and we just have to call it once. Then just one single line will be repeated instead of the whole block.